### PR TITLE
[SPARK-49306][PYTHON][SQL] Create DataFrame API support for new 'zeroifnull' and 'nullifzero' SQL functions

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

In https://github.com/apache/spark/pull/47817 we added new SQL functions `zeroifnull` and `nullifzero`.

In this PR we add Scala and Python DataFrame API endpoints for them.

For example, in Scala:

```
var df = Seq((0)).toDF("a")
df.selectExpr("nullifzero(0)").collect()
> null
df.select(nullifzero(lit(0))).collect()
> null

df.selectExpr("nullifzero(a)").collect()
> null
df.select(nullifzero(lit(5))).collect()
> 5

df = Seq[(Integer)]((null)).toDF("a")
df.selectExpr("zeroifnull(null)").collect()
> 5
df.select(nullifzero(lit(null))).collect()
> 0

df.selectExpr("zeroifnull(a)").collect()
> 0
df.select(zeroifnull(lit(5)))
> 5
```

### Why are the changes needed?

This improves DataFrame parity with the SQL API.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds unit test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @HyukjinKwon @MaxGekk

File: python/pyspark/sql/functions/builtin.py [hunk 1]
```diff
@@ -20681,6 +20681,26 @@ def nullif(col1: "ColumnOrName", col2: "ColumnOrName") -> Column:
...
+
+    Parameters
+    ----------
+    col : :class:`~pyspark.sql.Column` or str
+
+    Examples
+    --------
+    >>> df = spark.createDataFrame([(0,), (1,)], ["a"])
+    >>> df.select(nullifzero(df.a).alias('r')).collect()
+    [Row(r=None), Row(r=1)]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions/builtin.py:20699]: Could we add a test in the actual `test_functions.py`? Then the test will be automatically shared between Spark Connect and Spark Classic

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 2]
```diff
@@ -7866,6 +7874,14 @@ object functions {
    */
   def nvl2(col1: Column, col2: Column, col3: Column): Column = Column.fn("nvl2", col1, col2, col3)
 
+  /**
+   * Returns zero if `col` is null, or `col` otherwise.
+   *
+   * @group conditional_funcs
+   * @since 4.0.0
+   */
+  def zeroifnull(col: Column): Column = Column.fn("zeroifnull", col)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:7883]: Would need to implement this for Spark Connect in Scala too (e.g., `connector/connect/client/.../functions.scala`)

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 3]
```diff
@@ -331,6 +331,15 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
     checkAnswer(df.select(nullif(lit(5), lit(5))), Seq(Row(null)))
   }
 
+  test("nullifzero function") {
+    val df = Seq((0)).toDF("a")
+    checkAnswer(df.selectExpr("nullifzero(0)"), Seq(Row(null)))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:336]: You have already tested the SQL function in the previous PR. I do believe this check is not necessary.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 4]
```diff
@@ -331,6 +331,15 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
...
+    checkAnswer(df.select(nullifzero(lit(0))), Seq(Row(null)))
+
+    checkAnswer(df.selectExpr("nullifzero(a)"), Seq(Row(null)))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:339]: Let's skip this and add more tests for:
1. Different types: maps, arrays, dates and so on
2. Not-nullable column
3. Types support ordering and do not support 
4. Non-foldable column

File: python/pyspark/sql/functions/builtin.py [hunk 5]
```diff
@@ -20724,6 +20744,26 @@ def nvl2(col1: "ColumnOrName", col2: "ColumnOrName", col3: "ColumnOrName") -> Co
...
+    .. versionadded:: 4.0.0
+
+    Parameters
+    ----------
+    col : :class:`~pyspark.sql.Column` or str
+
+    Examples
+    --------
+    >>> df = spark.createDataFrame([(None,), (1,)], ["a"])
+    >>> df.select(zeroifnull(df.a).alias('r')).collect()
```
Comment (allisonwang-db) [inline python/pyspark/sql/functions/builtin.py:20761]: super nit:  `df.show()` is more user-friendly :)

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 3 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: sounds good, removed.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 4 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: sounds good, this is done.

**Review by dtenedor — COMMENTED** [review]
Thanks @MaxGekk @HyukjinKwon @allisonwang-db for your reviews, responded to your comments, please take another look.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 6]
```diff
@@ -331,6 +331,54 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
     checkAnswer(df.select(nullif(lit(5), lit(5))), Seq(Row(null)))
   }
 
+  test("nullifzero function") {
+    withTable("t") {
+      // Here we exercise a non-nullable, non-foldable column.
+      sql("create table t(col int not null) using csv")
+      sql("insert into t values (0)")
+      val df = sql("select col from t")
+      checkAnswer(df.selectExpr("nullifzero(col)"), Seq(Row(null)))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:340]: @dtenedor I do understand that the functions from Scala API call SQL functions, but could you test directly the added API.

Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Sounds good, done.

File: python/pyspark/sql/functions/builtin.py [hunk 7]
```diff
@@ -20681,6 +20681,31 @@ def nullif(col1: "ColumnOrName", col2: "ColumnOrName") -> Column:
...
+    col : :class:`~pyspark.sql.Column` or str
+
+    Examples
+    --------
+    >>> df = spark.createDataFrame([(0,), (1,)], ["a"])
+    >>> df.select(nullifzero(df.a).alias("result")).show()
+    +------+
+    |result|
+    +------+
+    |  None|
```
Comment (MaxGekk) [inline python/pyspark/sql/functions/builtin.py:20702]: Why do you expect `None` if the function `nullifzero()` should return `NULL` for `0`.

Comment (dtenedor) [reply python/pyspark/sql/functions/builtin.py]: Fixed this; it was indeed a typo and should say `NULL` instead of `None`.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to mater.
Thank you, @dtenedor and @HyukjinKwon for review.

**Pull request closed without merging**